pkg/instance/service.go

```diff
@@ -184,8 +186,60 @@ func (s service) unlink(id uint) error {
 	return s.instanceRepository.Unlink(instance)
 }
 
+func (s service) matchRequiredParameters(stackParams []model.StackRequiredParameter, instanceParams []model.InstanceRequiredParameter) []string {
+	unmatchedParams := make([]string, 0)
+	paramNames := make(map[string]struct{})
+
+	for _, param := range stackParams {
+		paramNames[param.Name] = struct{}{}
+	}
+
+	for _, param := range instanceParams {
+		if _, ok := paramNames[param.StackRequiredParameterID]; !ok {
+			unmatchedParams = append(unmatchedParams, param.StackRequiredParameterID)
+		}
+	}
+
+	return unmatchedParams
+}
+
+func (s service) matchOptionalParameters(stackParams []model.StackOptionalParameter, instanceParams []model.InstanceOptionalParameter) []string {
+	unmatchedParams := make([]string, 0)
+	paramNames := make(map[string]struct{})
+
+	for _, param := range stackParams {
+		paramNames[param.Name] = struct{}{}
+	}
+
+	for _, param := range instanceParams {
+		if _, ok := paramNames[param.StackOptionalParameterID]; !ok {
+			unmatchedParams = append(unmatchedParams, param.StackOptionalParameterID)
+		}
+	}
+
+	return unmatchedParams
+}
+
 func (s service) Save(instance *model.Instance) (*model.Instance, error) {
-	err := s.instanceRepository.Save(instance)
+	instanceStack, err := s.stackService.Find(instance.StackName)
+	if err != nil {
+		return nil, err
+	}
+
+	unmatchedReqParams := s.matchRequiredParameters(instanceStack.RequiredParameters, instance.RequiredParameters)
```

Comment:
To make `save` more readable you can put all the unmatched params logic into one function so this here reads like

```go

err:= validateParameters(instanceStack, instance)
if err != nil {
  return err
}
```